### PR TITLE
Create endpoint for linking nodes

### PR DESCRIPTION
**Pull request by anhduc130**

### CODE REVIEW

**Comment by wynnset** [comment]
Please also rebase your branch.

File: controller/class.tapestry-controller.php [hunk 1]
```diff
@@ -43,6 +43,10 @@ class TapestryController
         'INVALID_CHILD_NODE' => [
             'MESSAGE'   => 'Node is not a child of the tapestry',
             'STATUS'    => ['status' => 400]
+        ],
+        'INVALID_TAPESTRY' => [
```
Comment (wynnset) [inline controller/class.tapestry-controller.php]: The key here is not descriptive of the error. Maybe `INVALID_NEW_TAPESTRY` would be better.

File: controller/class.tapestry-controller.php [hunk 2]
```diff
@@ -59,16 +63,20 @@ public function __construct($postId = 0)
...
      */
-    public function updateTapestry($tapestry)
+    public function addTapestry($tapestry)
     {
+        if (isset($tapestry->links) || count($tapestry->nodes) > 1) {
+            return $this->_throwsError('INVALID_TAPESTRY');
+        }
         if (!$this->postId) {
             $this->postId = $this->_updatePost($tapestry, 'tapestry');
         }
```
Comment (wynnset) [inline controller/class.tapestry-controller.php]: What about the case that there is already a tapestry saved with the passed in ID? You would be adding _additional_ nodes, which is unexpected. Maybe you should check to make sure the tapestry is empty if `$this->postId` is passed in, otherwise return an error.

File: controller/class.tapestry-controller.php [hunk 3]
```diff
@@ -333,32 +339,23 @@ private function _isValidTapestryGroup($groupMetaId)
...
-            }
+            $nodePostId = $this->_updatePost($node, 'tapestry_node');
+            $metadata = $this->_makeMetadata($node, $nodePostId);
+            $node->id = add_post_meta($this->postId, 'tapestry_node', $metadata);
             update_post_meta($nodePostId, 'tapestry_node_data', $node);
         }
     }
 
-    private function _updateGroups($groups)
+    private function _addGroups($groups)
```
Comment (wynnset) [inline controller/class.tapestry-controller.php]: It would look cleaner if you change this function to `_addGroup` and accept a single group instead and do the loop over groups in the caller when needed.

File: controller/class.tapestry-controller.php (hunk 2 above)
Comment (anhduc130) [reply controller/class.tapestry-controller.php]: Yeah, I just realize that there will be no such scenario that `$this->postId` would be set when this method is called (the function in `endpoints.php` didn't initialize the postId when calling this method). I'd say it's safe to remove this check.

Comment (wynnset) [reply controller/class.tapestry-controller.php]: Then, I suggest adding a check to make sure `$this->postId` is not passed in.

File: controller/class.tapestry-controller.php [hunk 4]
```diff
@@ -43,54 +43,86 @@ class TapestryController
...
             if (!isset($node->permissions)) {
                 $node->permissions = (object)self::NODE_PERMISSIONS['DEFAULT'];
             }
         }
 
-        $this->_updateNodes($tapestry->nodes);
-        $this->_updateGroups($tapestry->groups);
+        foreach ($tapestry->nodes as $node) {
+            if ($this->_isValidTapestryNode($node->id)) {
+                return $this->_throwsError('NODE_ALREADY_EXISTS');
```
Comment (wynnset) [inline controller/class.tapestry-controller.php]: It's not a good idea to return an error after you have made changes to the database. I suggest moving the validation for both nodes and groups up before you call `_updatePost`

File: controller/class.tapestry-controller.php [hunk 5]
```diff
@@ -43,54 +43,86 @@ class TapestryController
...
+            }
+            $this->_addNode($node);
+        }
+
+        foreach ($tapestry->groups as $group) {
+            if ($this->_isValidTapestryGroup($group->id)) {
+                return $this->_throwsError('GROUP_ALREADY_EXISTS');
```
Comment (wynnset) [inline controller/class.tapestry-controller.php]: As per the previous comment, move the validation up.

File: controller/class.tapestry-controller.php [hunk 6]
```diff
@@ -43,54 +43,86 @@ class TapestryController
...
+            'STATUS'    => ['status' => 500]
         ]
     ];
     private $postId;
 
     /**
      * Constructor
      */
-    public function __construct($postId = 0)
+    public function __construct($postId)
```
Comment (wynnset) [inline controller/class.tapestry-controller.php]: Why did you remove the default? This means this argument will be mandatory, so even new tapestries will need to pass in 0 to work.

File: controller/class.tapestry-controller.php [hunk 7]
```diff
@@ -43,54 +43,86 @@ class TapestryController
...
         }
         if (!isset($tapestry->nodes)) {
             $tapestry->nodes = [];
         }
         if (!isset($tapestry->groups)) {
             $tapestry->groups = [];
         }
+
+        $this->postId = $this->_updatePost($tapestry, 'tapestry');
+
```
Comment (wynnset) [inline controller/class.tapestry-controller.php]: I suggest you dedicate this function solely to adding a new tapestry, not its nodes, groups, etc.

File: controller/class.tapestry-controller.php (hunk 6 above)
Comment (anhduc130) [reply controller/class.tapestry-controller.php]: What I am trying to do is that if there is no `$postId` passed in or `$postId` is valid, I would set it to NULL. The methods will then check for the $postId.

File: controller/class.tapestry-controller.php (hunk 4 above)
Comment (anhduc130) [reply controller/class.tapestry-controller.php]: Yeah. I guess we are going to remove this anyway.

File: controller/class.tapestry-controller.php (hunk 6 above)
Comment (anhduc130) [reply controller/class.tapestry-controller.php]: Set the parameter default to NULL.

File: controller/class.tapestry-controller.php [hunk 8]
```diff
@@ -43,56 +43,76 @@ class TapestryController
...
+            'STATUS'    => ['status' => 400]
+        ],
+        'LINKS_EXIST_IN_NEW_TAPESTRY' => [
+            'MESSAGE'   => 'Links exist when creating a new Tapestry',
+            'STATUS'    => ['status' => 400]
+        ],
+        'POST_ID_ALREADY_SET' => [
+            'MESSAGE'   => 'PostID is set when creating a new Tapestry',
+            'STATUS'    => ['status' => 500]
         ]
```
Comment (wynnset) [inline controller/class.tapestry-controller.php]: Could you please phrase your error messages to tell them what needs to be changed for their request to be successful? So "Settings are required to created a new Tapestry", "Nodes should not be passed in when creating a new Tapestry", etc.

File: controller/class.tapestry-controller.php [hunk 9]
```diff
@@ -43,56 +43,76 @@ class TapestryController
...
     ];
     private $postId;
 
     /**
      * Constructor
      */
-    public function __construct($postId = 0)
+    public function __construct($postId = null)
```
Comment (wynnset) [inline controller/class.tapestry-controller.php]: Documentation for [wp_insert_post](https://developer.wordpress.org/reference/functions/wp_insert_post/) states that it requires postId to be 0 to create a new post. I suggest you change this to 0 and also do the same below since you may be passing this to the wp_insert_post function at some point.

File: controller/class.tapestry-controller.php [hunk 10]
```diff
@@ -43,56 +43,76 @@ class TapestryController
...
     {
-        if (!$this->postId) {
-            $this->postId = $this->_updatePost($tapestry, 'tapestry');
+        if ($this->postId) {
+            return $this->_throwsError('POST_ID_ALREADY_SET');
         }
-        if (!isset($tapestry->nodes)) {
-            $tapestry->nodes = [];
+        if (empty($tapestry->settings)) {
+            return $this->_throwsError('INVALID_NEW_TAPESTRY');
```
Comment (wynnset) [inline controller/class.tapestry-controller.php]: Again, I suggest you change this error index to be more descriptive like `'SETTINGS_MISSING_IN_NEW_TAPESTRY'` or something

File: controller/class.tapestry-controller.php [hunk 11]
```diff
@@ -211,7 +235,36 @@ public function updateTapestrySettings($settings)
...
+    }
+
+    /**
+     * Add A Tapestry Link
+     * 
+     * @param   Object  $link   Tapestry link
+     * 
+     * @return  Object  $link 
+     */
+    public function addTapestryLink($link)
```
Comment (wynnset) [inline controller/class.tapestry-controller.php]: Could you please move this function before the `updateTapestrySettings` function so it's a bit more organized? All the add functions should be beside eachother as much as possible.

File: controller/class.tapestry-controller.php (hunk 8 above)
Comment (wynnset) [reply controller/class.tapestry-controller.php]: Please update the message for the last item `'POST_ID_ALREADY_SET'` as well.
